### PR TITLE
Removes types from SearchRequest and QueryShardContext

### PR DESCRIPTION
**Pull request by colings86**
This change will remove the notion of types completely from the Search API internals.

For classes that are serialised (such as `SearchRequest`, `IdsQuery`, `ShardLocalSearchRequest`) we need to check for the wire version and deal with the types serialisation if the other node is `<8.0.0`

### CODE REVIEW

**Comment by colings86** [comment]
@jpountz I was hoping this would be easier to split into incremental changes. All I managed to do was split the changes from removing types from `SearchRequest` into a separate commit to the ones removing types from `QueryShardContext`. I can split the commits into 2 separate PRs if you prefer? Also open to thoughts on better ways to split this if think it needs to be made smaller?

File: server/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 1]
```diff
@@ -191,7 +189,10 @@ public SearchRequest(StreamInput in) throws IOException {
         preference = in.readOptionalString();
         scroll = in.readOptionalWriteable(Scroll::new);
         source = in.readOptionalWriteable(SearchSourceBuilder::new);
-        types = in.readStringArray();
+        if (in.getVersion().before(Version.V_8_0_0)) {
+            // types no longer relevant so ignore
+            in.readStringArray();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/action/search/SearchRequest.java]: should we fail if the array is not empty?

File: server/src/test/java/org/elasticsearch/search/aggregations/pipeline/SerialDiffIT.java [hunk 2]
```diff
@@ -275,7 +275,7 @@ public void testBasicDiff() {
     public void testInvalidLagSize() {
         try {
             client()
-                .prepareSearch("idx").setTypes("type")
+                    .prepareSearch("idx")
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/aggregations/pipeline/SerialDiffIT.java]: it should be indented on the same level as `.addAggregation`?

**Review by jpountz — CHANGES_REQUESTED** [review]
It looks good overall. I think we should fail deserialization of request builders when the set of types is not empty? My worry is that we might hide bugs otherwise, eg. if a user is querying a different type than the type of the index: if we ignore types, the query might match some documents while it shouldn't?

**Comment by colings86** [comment]
@jpountz thanks for the review, I pushed a commit to address your comments

**Comment by colings86** [comment]
@elasticmachine run elasticsearch-ci/1

**Review by jpountz — APPROVED** [review]

**Pull request merged**